fontforge/splineutil.c

```diff
@@ -7353,6 +7353,52 @@ int SplineSetsRemoveAnnoyingExtrema(SplineSet *ss,bigreal err) {
 return( changed );
 }
 
+int SplineRemoveWildControlPoints(Spline *s, bigreal distratio) {
+	// If the distance between the control point and its base
+	// exceeds the the distance between the base points 
+	// or the control points by a great factor,
+	// it is likely erroneous and is likely to cause problems.
+	// So we remove it.
+	if (s->from == NULL || s->to == NULL)
+		return 0;
+	int changed;
+	bigreal pdisplacement = DistanceBetweenPoints(&s->from->me, &s->to->me);
+	bigreal bcdisplacement0 = 0.0;
+	bigreal bcdisplacement1 = 0.0;
+	bigreal ccdisplacement = 0.0;
+	if (!s->from->nonextcp)
+		bcdisplacement0 = DistanceBetweenPoints(&s->from->me, &s->from->nextcp);
+	if (!s->to->noprevcp)
+		bcdisplacement1 = DistanceBetweenPoints(&s->to->me, &s->to->prevcp);
+	if (!s->from->nonextcp && !s->to->noprevcp)
+		ccdisplacement = DistanceBetweenPoints(&s->from->nextcp, &s->to->prevcp);
+	bigreal basis = MIN(pdisplacement, ccdisplacement);
+	if (basis == 0 || MAX(bcdisplacement0, bcdisplacement1) / pdisplacement > basis) {
```

Comment:
Maybe this is something specific to the algorithm feeding into this check, but why should ccdisplacement==0 (via basis==0) be significant here? Consider a cubic curve with points in an equilateral triangle, with `from` and `to` *d* units apart `nextcp` and `prevcp` co-located *d* units part from each of `from` and `to`. Why convert to a line in that case?

Reply:
Oh. This logic is wrong. `basis` was supposed to be the denominator and `distratio` the other side of the comparison. I made a bunch of changes that definitely merited a full review afterwards but got distracted by the great test results.

Reply:
The `basis == 0` check makes more sense if its going to be the denominator (I was wondering about that). But the other comment still applies -- even if `ccdisplacement` is going to figure into the denominator I don't get the zero check for it.

Reply:
Ah, yes. `basis` ought to come from `MAX` rather than from `MIN`. These curves are most problematic when they are really sharp. The case I was trying to accommodate (by a rough heuristic that would let distant collinear control points slide) was a single-spline loop starting and ending at the same point with divergent control points.